### ISSUE TITLE
Research Spike: Swift cannot use grpc in the background

### ISSUE THREAD

**Issue by gmale** [zcash/lightwalletd#305]
This ticket is to research what it would take to enable RESTful JSON services from lightwalletd. Can we identify the lowest-friction approach to turning on JSON? Supposedly, it's fairly simple to have a Go server provide both _gRPC over http2_ and _Json over http/1.x_.

Time box:
1 work day

**Comment by LarryRuane** [comment]
@benzcash, can you provide any insight about how to go about this? I'd be glad to make the changes but I probably don't have the necessary base knowledge. Thank you!

**Comment by benzcash** [comment]
I looked into it a bit, seems straight forward https://github.com/grpc-ecosystem/grpc-gateway.

There are lots of articles about combining both on the saver server port (which would be nice) . For example https://www.clarifai.com/blog/muxing-together-grpc-and-http-traffic-with-grpc-gateway

**Comment by gmale** [comment]
I wonder how caching would work,  once we have HTTP/1.x RESTful services working for compact blocks:
- do we just leverage the existing on-disk cache and just serialize everything to JSON on the fly?
- do we cache actual JSON responses (probably not feasible because the wallets will request a variety of ranges)

**Comment by lindanlee** [comment]
Pause until pacu confirms that he needs this support. We got some new information today that may change our requirements.

**Comment by LarryRuane** [comment]
Just wondering, would it be useful to make the entire API available over HTTP even if it turns out that swift doesn't need it for this purpose? And just so I'm clear, a way to demonstrate that the goal has been met would be if all of the production (non-darkside) functionality could be accessed using `curl`? The gRPCs that use streaming for either arguments or returns could be a bit of a problem.

**Comment by gmale** [comment]
Generally, I'm mildly concerned about the can of worms that could open: for instance someone DoS-ing the server by requesting all blocks, since sapling activation and getting a giant JSON response. Perhaps those types of things can be mitigated by actively validating the request parameters and denying abusive requests.

I imagine this is more of a Brad question but the answer would probably depend on the level of effort involved. If we can get RESTful JSON safely, and it's a low effort to do so and **_maintain it_**, then it seems like a no-brainer. If it become a maintenance or security headache, or it takes many sprints to implement, then that seems like a non-starter.

(NOTE: I believe gRPC **_is HTTP_**--effectively just an implementation HTTP/2 so I'm speaking in terms of "RESTful JSON" over HTTP/1.2)

**Comment by gmale** [comment]
TL;DR: IMHO, it "would be useful" if it's easy to implement and maintain.

**Comment by benzcash** [comment]
It should be easy with the gateway.

From the README
> This project aims to provide that HTTP+JSON interface to your gRPC service. A small amount of configuration in your service to attach HTTP semantics is all that's needed to generate a reverse-proxy with this library.

I would be interested in helping since this would help with monitoring of the service :grin:
